File: samples/legacy-architecture-samples/simple-stack-example-mvvm-fragments/src/main/java/com/zhuinden/simplestackexamplemvvm/core/database/BaseDao.java
package com.zhuinden.simplestackexamplemvvm.core.database;

import androidx.annotation.Nullable;

import java.util.Collections;
import java.util.List;

public abstract class BaseDao<T> implements Dao<T> {
    private final DatabaseManager databaseManager;
    private final DatabaseManager.Table table;
    private final DatabaseManager.Mapper<T> mapper;

    public BaseDao(DatabaseManager databaseManager, DatabaseManager.Table table, DatabaseManager.Mapper<T> mapper) {
        this.databaseManager = databaseManager;
        this.table = table;
        this.mapper = mapper;
    }

    @Override
    public final DatabaseManager.Table getTable() {
        return table;
    }

    @Override
    public void refresh() {
        databaseManager.refresh(table);
    }

    @Override
    public final T findOne(@Nullable String id) {
        if(id == null) {
            return databaseManager.findOne(table, mapper, QueryBuilder.of(table)
                    .where("1 = 0")
                    .buildDefinition());
        } else {
            return databaseManager.findOne(table, mapper, QueryBuilder.of(table)
                    .where(getTable().getIdFieldName() + " = ?", id)
                    .limit(1)
                    .buildDefinition());
        }
    }

    @Override
    public final T findOne(DatabaseManager.QueryDefinition queryDefinition) {
        return databaseManager.findOne(table, mapper, queryDefinition);
    }

    @Override
    public final List<T> findAll() {
        return databaseManager.findAll(table, mapper);
    }

    @Override
    public final List<T> findAll(DatabaseManager.QueryDefinition queryDefinition) {
        return databaseManager.findAll(table, mapper, queryDefinition);
    }

    @Override
    public final void insert(T element) {
        databaseManager.insert(table, mapper, Collections.singletonList(element));
    }

    @Override
    public final void insert(List<T> list) {
        databaseManager.insert(table, mapper, list);
    }

    @Override
    public final void delete(T element) {
        databaseManager.delete(table, Collections.singletonList(element));
    }

    @Override
    public final void delete(List<T> list) {
        databaseManager.delete(table, list);
    }

    @Override
    public final void deleteAll() {
        databaseManager.deleteAll(table);
    }

    @Override
    public final LiveResults<T> findAllWithChanges() {
        return databaseManager.findAllWithChanges(table, mapper, QueryBuilder.of(table).buildDefinition());
    }

    @Override
    public final LiveResults<T> findAllWithChanges(DatabaseManager.QueryDefinition queryDefinition) {
        return databaseManager.findAllWithChanges(table, mapper, queryDefinition);
    }
}

File: samples/advanced-samples/mvvm-sample/src/main/java/com/zhuinden/simplestackexamplemvvm/core/database/BaseDao.kt
package com.zhuinden.simplestackexamplemvvm.core.database


import com.zhuinden.simplestackexamplemvvm.core.database.DatabaseManager.QueryDefinition

abstract class BaseDao<T>(
    private val databaseManager: DatabaseManager,
    override val table: DatabaseManager.Table,
    private val mapper: DatabaseManager.Mapper<T>
) : Dao<T> {
    override fun refresh() {
        databaseManager.refresh(table)
    }

    override fun findOne(id: String): T? {
        return if (id == null) {
            databaseManager.findOne(table, mapper, QueryBuilder.of(table)
                .where("1 = 0")
                .buildDefinition())
        } else {
            databaseManager.findOne(table, mapper, QueryBuilder.of(table)
                .where(table.idFieldName + " = ?", id)
                .limit(1)
                .buildDefinition())
        }
    }

    override fun findOne(queryDefinition: QueryDefinition): T? {
        return databaseManager.findOne(table, mapper, queryDefinition)
    }

    override fun findAll(): List<T> {
        return databaseManager.findAll(table, mapper)
    }

    override fun findAll(queryDefinition: QueryDefinition): List<T> {
        return databaseManager.findAll(table, mapper, queryDefinition)
    }

    override fun insert(element: T) {
        databaseManager.insert(table, mapper, listOf(element))
    }

    override fun insert(list: List<T>) {
        databaseManager.insert(table, mapper, list)
    }

    override fun delete(element: T) {
        databaseManager.delete(table, element)
    }

    override fun deleteList(list: List<T>) {
        databaseManager.deleteList(table, list)
    }

    override fun deleteAll() {
        databaseManager.deleteAll(table)
    }

    override fun findAllWithChanges(): LiveResults<T> {
        return databaseManager.findAllWithChanges(table, mapper, QueryBuilder.of(table).buildDefinition())
    }

    override fun findAllWithChanges(queryDefinition: QueryDefinition): LiveResults<T> {
        return databaseManager.findAllWithChanges(table, mapper, queryDefinition)
    }
}